docs/design/kademlia-removal.md

```diff
@@ -0,0 +1,84 @@
+# Kademlia Removal
+
+## Abstract
+
+This design document outlines the communication protocol between satellites and 
+storage nodes, network refreshes, and kademlia removal given the satellite opt-in 
+capability for storage nodes. 
+
+## Background
+User research on how Storage Node Operators would connect to satellites resulted in a 
+simple solution called Opt-in SNO select. Satellites wait for storage nodes to work 
+with them, and storage nodes manually select the satellites they want to work with. 
+The initial implementation of this allows SNOs to update their trusted satellite list 
+in a hardcoded configuration file, but future improvements will enable users to manage 
+this list through a web console. Due to the nature of this change, the Storj network 
+at this point in time no longer requires the Kademlia DHT or related entities. 
+We will replace this protocol with direct communication between satellites and 
+storage nodes, and keep the network fresh without kademlia node discovery and random 
+lookups.
+
+## Design
+
+1. Nodes reach out to satellites that they want to work with
+    
+    - At the time of writing, a storage node operator can input a list of trusted satellite 
+    IDs and addresses into their configuration file on setup. There are several tardigrade-level 
+    satellites already listed as a preset. This list is imported into a Go map data structure 
+    within the “trust” package.
+    - Eventually during the design and implementation of trusted satellite management, 
+    this configured list may be imported into a on-disk data store, eg SQL table. Users will 
+    be able to modify this list through the storage node operator console. However, this is a 
+    separate discussion from this design document. This design is satisfied with a static trusted 
+    satellites list.
+    - Nodes will be able to communicate with satellites through the “transport” and “trust” packages 
+    directly rather than using kademlia to traverse the network to find the address of a given ID.
+
+2. Network refreshes at a regular interval
+    - Nodes will keep themselves up to date in the network by pinging all the satellites in their 
```

Comment:
that's why it makes sense to avoid SSL earlier. nonetheless, with kademlia satellites are doing this much work if not more

Reply:
The satellite can loadbalance the pings. With 1M storage nodes the satellite might take a long time for one ping itteration. We might get issues because the interval is too high but the satellite will not get DDoS.

If we turn that arround that means all storage nodes will ping the satellite. If we add a nice little bug that all storage nodes are crashing at the 29th feb they will restart all at the same time. They will take the satellite offline and because they failed to contact the satellite they will retry. How do we protect the satellite in that situation?

Reply:
they won't take the satellite offline if they just send single udp packets. the satellite can ignore them. the satellite will only go down if it chooses to engage

Reply:
What if the packet is dropped and the SN's uptime suffers?

Reply:
yeah that would stink. i think team green is planning on some kind of reaper service that checks in with storage nodes we haven't heard from recently

Reply:
If we're using udp, are we going to use DTLS?

Reply:
no, but littleskunk points out the udp message should at least be signed. let's avoid DTLS though

Reply:
I thought that this was basically the reaper service!

Reply:
Should explicitly state that they do this on startup (to reduce the amount of time they appear offline)?

Reply:
that's a good idea, yeah

Reply:
@wthorp @jtolds what do I need to add here?

Reply:
1) storage nodes should notify satellites when they start up, wait a random amount of time (to add jitter http://highscalability.com/blog/2012/4/17/youtube-strategy-adding-jitter-isnt-a-bug.html), then start reporting in roughly on the hour
2) when they report in to the satellite's new-node-notification tipbox, it can be over ssl, but it doesn't have to be. if it isn't over ssl it should at least be signed.
3) there should be a reaper service that checks out what's going on with nodes that haven't reported in in a while

Reply:
that's just my summary of the comment thread, not where it should go in the design doc

Reply:
Is this design doc intended to apply to just tardigrade satellites, or others? If others, I suggest that satellites can determine how often nodes are supposed to check in for uptime checks. Perhaps there is an "introduction ping" that occurs the first time the node comes online, and when the satellite responds, it includes how often nodes are expected to check back in to maintain good reputation.

Reply:
this is supposed to apply to others, but i'd like to scope configurable windows for a later feature and we don't tackle that now